micrometer/src/main/java/feign/micrometer/MeteredDecoder.java

```diff
@@ -46,41 +49,81 @@ public MeteredDecoder(Decoder decoder, MeterRegistry meterRegistry, MetricName m
-          .count();
+      sample.stop(createExceptionTimer(response, type, e));
+      createExceptionCounter(response, type, e).count();
       throw new IOException(e);
     }
 
-    body.ifPresent(b -> meterRegistry.summary(
-        metricName.name("response_size"),
-        metricName.tag(template.methodMetadata(), template.feignTarget())).record(b.count()));
+    body.ifPresent(b -> createSummary(response, type).record(b.count()));
 
     return decoded;
   }
 
+  private Timer createSuccessTimer(Response response, Type type) {
+    final List<Tag> successTags = extraSuccessTags(response, type);
+    final Tag[] tags = successTags.toArray(new Tag[] {});
+    final RequestTemplate template = response.request().requestTemplate();
+    final Tags allTags = metricName.tag(template.methodMetadata(), template.feignTarget(), tags);
+    return meterRegistry.timer(metricName.name(), allTags);
+  }
+
+  private Timer createExceptionTimer(Response response, Type type, Exception exception) {
+    final List<Tag> exceptionTags = extraExceptionTags(response, type, exception);
+    final Tag[] tags = exceptionTags.toArray(new Tag[] {});
+    final RequestTemplate template = response.request().requestTemplate();
+    final Tags allTags = metricName.tag(template.methodMetadata(), template.feignTarget(), tags);
+    return meterRegistry.timer(metricName.name("exception"), allTags);
+  }
+
+  private Counter createExceptionCounter(Response response, Type type, Exception exception) {
+    final List<Tag> exceptionTags = extraExceptionTags(response, type, exception);
+    final Tag[] tags = exceptionTags.toArray(new Tag[] {});
+    final RequestTemplate template = response.request().requestTemplate();
+    final Tags allTags = metricName.tag(template.methodMetadata(), template.feignTarget(), tags);
+    return meterRegistry.counter(metricName.name("error_count"), allTags);
+  }
+
+  private DistributionSummary createSummary(Response response, Type type) {
+    final List<Tag> successTags = extraSummaryTags(response, type);
+    final Tag[] tags = successTags.toArray(new Tag[] {});
+    final RequestTemplate template = response.request().requestTemplate();
+    final Tags allTags = metricName.tag(template.methodMetadata(), template.feignTarget(), tags);
+    return meterRegistry.summary(metricName.name("response_size"), allTags);
+  }
+
+  protected List<Tag> extraSuccessTags(Response response, Type type) {
+    return Collections.emptyList();
+  }
+
+  protected List<Tag> extraExceptionTags(Response response, Type type, Exception exception) {
+    final RequestTemplate template = response.request().requestTemplate();
+    final List<Tag> result = new ArrayList<>();
+    result.add(Tag.of("uri", template.path()));
+    result.add(Tag.of("exception_name", exception.getClass().getSimpleName()));
+    return result;
+  }
+
+  protected List<Tag> extraSummaryTags(Response response, Type type) {
```

Comment:
I assume you are using this for some internal metrics (I would be curious to know what you are going =] )

I would prefer to see code to live in the `MetricName` interface... that would give a single place for all your tag customization needs.

We would have a few methods, one called `requestExtraTags` other one `responseExtraTags`, another like `exceptionExtraTags`

What do you think?

Reply:
I extracted tags resolution to a separate extendable class. However, seems like depending on where we create tags (i.e. client, encoder/decoder, method handler) we have really different parameters to build tags upon. I added defaultTags method, which might be helpful for customization in general.